carla/recourse_methods/catalog/growing_spheres/library/gs_counterfactuals.py

```diff
@@ -0,0 +1,159 @@
+    # Thibaut Laugel et al (2018), "Comparison-based Inverse Classification for
+    # Interpretability in Machine Learning" -- International Conference on Information Processing
+    # and Management of Uncertainty in Knowledge-Based Systems (2018)
+
+    """
+    :param n_search_samples: int > 0
+    :param instance: numpy input point array
+    :param high: float>= 0, h>l; upper bound
+    :param low: float>= 0, l<h; lower bound
+    :param p: float>= 1; norm
+    :return: candidate counterfactuals & distances
+    """
+
+    delta_instance = np.random.randn(n_search_samples, instance.shape[1])
+    dist = np.random.rand(n_search_samples) * (high - low) + low  # length range [l, h)
+    norm_p = LA.norm(delta_instance, ord=p_norm, axis=1)
+    d_norm = np.divide(dist, norm_p).reshape(-1, 1)  # rescale/normalize factor
+    delta_instance = np.multiply(delta_instance, d_norm)
+    candidate_counterfactuals = instance + delta_instance
+
+    return candidate_counterfactuals, dist
+
+
+def growing_spheres_search(
+    instance,
+    keys_mutable,
+    keys_immutable,
+    continuous_cols,
+    binary_cols,
+    feature_order,
+    model,
+    n_search_samples=1000,
+    p_norm=2,
+    step=0.2,
+    max_iter=1000,
+):
+
+    """
+    :param instance: df
+    :param step: float > 0; step_size for growing spheres
+    :param n_search_samples: int > 0
+    :param model: sklearn classifier object
+    :param p_norm: float=>1; denotes the norm (classical: 1 or 2)
+    :param max_iter: int > 0; maximum # iterations
+    :param keys_mutable: list; list of input names we can search over
+    :param keys_immutable: list; list of input names that may not be searched over
+    :return:
+    """  #
+
+    # correct order of names
+    keys_correct = feature_order
+    # divide up keys
+    keys_mutable_continuous = list(set(keys_mutable) - set(binary_cols))
+    keys_mutable_binary = list(set(keys_mutable) - set(continuous_cols))
+
+    # Divide data in 'mutable' and 'non-mutable'
+    # In particular, divide data in 'mutable & binary' and 'mutable and continuous'
+    instance_immutable_replicated = np.repeat(
+        instance[keys_immutable].values.reshape(1, -1), n_search_samples, axis=0
+    )
+    instance_replicated = np.repeat(
+        instance.values.reshape(1, -1), n_search_samples, axis=0
+    )
+    instance_mutable_replicated_continuous = np.repeat(
+        instance[keys_mutable_continuous].values.reshape(1, -1),
+        n_search_samples,
+        axis=0,
+    )
+    # instance_mutable_replicated_binary = np.repeat(
+    #     instance[keys_mutable_binary].values.reshape(1, -1), n_search_samples, axis=0
+    # )
+
+    # init step size for growing the sphere
+    low = 0
+    high = low + step
+
+    # counter
+    count = 0
+    counter_step = 1
+
+    # get predicted label of instance
+    instance_label = np.argmax(model.predict_proba(instance.values.reshape(1, -1)))
+
+    while True:
```

Comment:
`While True` have to be used with care and are rarely the right choice for what we want ot achieve. Instead we could use a flag, e.g. `counterfactuals_found` which we set to true further down in the code.

```
counterfactuals_found = False
while counterfactual_founds == False:
  ...
  
  if len(candidate_dist) > 0: # certain candidates generated
        min_index = np.argmin(candidate_dist)
        candidate_counterfactual_star = candidate_counterfactuals[min_index]
        counterfactuals_found = True
  
  
  # no candidate found & push search range outside
  low = high
  high = low + step

```

Also note that there is a "failsave" condition missing in case no counterfactual can ever be found. We could think e.g. when the lows / highs are outside of a certain range then we can abort the search process

Reply:
Ah right a totally overlooked this. To make it extra nice once could also include this condition in the while loop and have the default an empty array. (Totally no blocker though)

```
candidate_counterfactual_star[:] = np.nan
while not counterfactuals_found or count > max_iter:
        count = count + counter_step
        ...

return candidate_counterfactual_start
```